src/libstore/outputs-spec.hh

```diff
@@ -1,32 +1,91 @@
 #pragma once
 
+#include <optional>
+#include <set>
 #include <variant>
 
-#include "util.hh"
-
-#include "nlohmann/json_fwd.hpp"
+#include "json-impls.hh"
 
 namespace nix {
 
-typedef std::set<std::string> OutputNames;
+struct OutputNames : std::set<std::string> {
+    using std::set<std::string>::set;
+
+    // These need to be "inherited manually"
+    OutputNames(const std::set<std::string> & s)
+        : std::set<std::string>(s)
+    { }
+    OutputNames(std::set<std::string> && s)
+        : std::set<std::string>(s)
+    { }
 
-struct AllOutputs {
-    bool operator < (const AllOutputs & _) const { return false; }
+    /* This set should always be non-empty, so we delete this
+       constructor in order make creating empty ones by mistake harder.
+       */
+    OutputNames() = delete;
 };
 
-struct DefaultOutputs {
-    bool operator < (const DefaultOutputs & _) const { return false; }
+struct AllOutputs : std::monostate { };
+
+typedef std::variant<AllOutputs, OutputNames> _OutputsSpecRaw;
+
+struct OutputsSpec : _OutputsSpecRaw {
+    using Raw = _OutputsSpecRaw;
+    using Raw::Raw;
+
+    /* Force choosing a variant */
+    OutputsSpec() = delete;
+
+    using Names = OutputNames;
+    using All = AllOutputs;
+
+    inline const Raw & raw() const {
+        return static_cast<const Raw &>(*this);
+    }
+
+    inline Raw & raw() {
+        return static_cast<Raw &>(*this);
+    }
+
+    bool contains(const std::string & output) const;
+
+    /* Modify the receiver outputs spec so it is the union of it's old value
+       and the argument. Return whether the output spec needed to be modified
+       --- if it didn't it was already "large enough". */
+    bool merge(const OutputsSpec & outputs);
+
+    /* Parse a string of the form 'output1,...outputN' or
+       '*', returning the outputs spec. */
+    static OutputsSpec parse(std::string_view s);
+    static std::optional<OutputsSpec> parseOpt(std::string_view s);
+
+    std::string to_string() const;
 };
 
-typedef std::variant<DefaultOutputs, AllOutputs, OutputNames> OutputsSpec;
+struct DefaultOutputs : std::monostate { };
```

Comment:
same pattern here. I have seen many code bases where developers subclassed stl classes - none of them aged well.